File: excelium-sheets/src/main/java/excelium/sheets/SheetsWriter.java
/*
 * MIT License
 *
 * Copyright (c) 2018 Excelium
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package excelium.sheets;

import com.google.api.services.sheets.v4.Sheets;
import com.google.api.services.sheets.v4.model.BatchUpdateValuesRequest;
import com.google.api.services.sheets.v4.model.ValueRange;
import excelium.core.writer.AbstractTestWriter;

import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.*;

/**
 * Writes Google Sheet spreadsheets.
 *
 * @author PhungDucKien
 * @since 2018.05.01
 */
public class SheetsWriter extends AbstractTestWriter {

    /**
     * Flush interval. Once the number of changes is greater than this config, flushes changes.
     */
    private static final int FLUSH_INTERVAL = 100;

    /**
     * Spreadsheet id
     */
    private String spreadsheetId;

    /**
     * Sheets service
     */
    private Sheets sheetsService;

    /**
     * Instantiates a new Sheets writer.
     *
     * @param sheetsService the sheets service
     * @param spreadsheetId the spreadsheet id
     */
    SheetsWriter(Sheets sheetsService, String spreadsheetId) {
        this.spreadsheetId = spreadsheetId;
        this.sheetsService = sheetsService;
    }

    @Override
    public synchronized void flush() throws IOException {
        List<ValueRange> valueRanges = new ArrayList<>();
        for (String range : flushValues.keySet()) {
            ValueRange valueRange = new ValueRange().setRange(range).setValues(flushValues.get(range));
            valueRanges.add(valueRange);
        }

        BatchUpdateValuesRequest request = new BatchUpdateValuesRequest();
        request.setData(valueRanges);
        request.setValueInputOption("USER_ENTERED");
        sheetsService.spreadsheets().values().batchUpdate(spreadsheetId, request).execute();
        flushValues.clear();
    }

    @Override
    public void batchSetRangeCellValues(Map<String, List<List<Object>>> values, String... ranges) throws IOException {
        super.batchSetRangeCellValues(interpretInputData(values), ranges);
        if (flushValues != null && flushValues.size() >= FLUSH_INTERVAL) {
            flush();
        }
    }

    /**
     * Interpret input data to make sure Google Sheets understand it.
     *
     * @param values the input data
     * @return the interpreted data
     */
    private Map<String, List<List<Object>>> interpretInputData(Map<String, List<List<Object>>> values) {
        Map<String, List<List<Object>>> userEnterValues = new HashMap<>();
        for (String range : values.keySet()) {
            List<List<Object>> rangeValues = values.get(range);
            List<List<Object>> userEnterRangeValues = new ArrayList<>();
            for (List<Object> rowValues : rangeValues) {
                List<Object> userEnterRowValues = new ArrayList<>();
                for (Object colValue : rowValues) {
                    if (colValue instanceof String) {
                        userEnterRowValues.add(colValue);
                    } else if (colValue instanceof Boolean) {
                        userEnterRowValues.add(colValue);
                    } else if (colValue instanceof Date) {
                        SimpleDateFormat df = new SimpleDateFormat("yyyy/MM/dd");
                        userEnterRowValues.add(df.format(colValue));
                    } else if (colValue instanceof Double) {
                        userEnterRowValues.add(colValue);
                    } else if (colValue instanceof Integer) {
                        userEnterRowValues.add(Integer.valueOf((int) colValue).doubleValue());
                    } else if (colValue instanceof Long) {
                        userEnterRowValues.add(Long.valueOf((long) colValue).doubleValue());
                    } else if (colValue == null) {
                        userEnterRowValues.add("");
                    }
                }
                userEnterRangeValues.add(userEnterRowValues);
            }
            userEnterValues.put(range, userEnterRangeValues);
        }
        return userEnterValues;
    }
}

File: packages/sheets/src/SheetsWriter.ts
/*
 * Licensed to the Software Freedom Conservancy (SFC) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The SFC licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { AbstractTestWriter } from '@excelium/core';
import { sheets_v4 } from 'googleapis';
import moment from 'moment';

/**
 * Writes Google Sheet spreadsheets.
 *
 * @author PhungDucKien
 * @since 2018.05.01
 */
export default class SheetsWriter extends AbstractTestWriter {
  /**
   * Flush interval. Once the number of changes is greater than this config, flushes changes.
   */
  private static readonly FLUSH_INTERVAL = 100;

  /**
   * Sheets service
   */
  private sheetsService: sheets_v4.Sheets;

  /**
   * Spreadsheet id
   */
  private spreadsheetId: string;

  /**
   * Instantiates a new Sheets writer.
   *
   * @param sheetsService the sheets service
   * @param spreadsheetId the spreadsheet id
   */
  constructor(sheetsService: sheets_v4.Sheets, spreadsheetId: string) {
    super();
    this.spreadsheetId = spreadsheetId;
    this.sheetsService = sheetsService;
  }

  public async flush(): Promise<void> {
    const valueRanges = [] as sheets_v4.Schema$ValueRange[];
    this.flushValues.forEach((values, range) => {
      valueRanges.push({
        range,
        values,
      });
    });

    return this.sheetsService.spreadsheets.values
      .batchUpdate({
        spreadsheetId: this.spreadsheetId,
        requestBody: {
          data: valueRanges,
          valueInputOption: 'USER_ENTERED',
        },
      })
      .then(() => {
        this.flushValues.clear();
      });
  }

  public batchSetRangeCellValues(values: Map<string, any[][]>, ranges: string[]): void {
    super.batchSetRangeCellValues(this.interpretInputData(values), ranges);
    if (this.flushValues != null && this.flushValues.size >= SheetsWriter.FLUSH_INTERVAL) {
      this.flush();
    }
  }

  /**
   * Interpret input data to make sure Google Sheets understand it.
   *
   * @param values the input data
   * @return the interpreted data
   */
  private interpretInputData(values: Map<string, any[][]>): Map<string, any[][]> {
    const userEnterValues = new Map<string, any[][]>();
    values.forEach((rangeValues, range) => {
      const userEnterRangeValues = [] as any[][];
      for (const rowValues of rangeValues) {
        const userEnterRowValues = [] as any[];
        for (const colValue of rowValues) {
          if (typeof colValue === 'string' || colValue instanceof String) {
            userEnterRowValues.push(colValue);
          } else if (typeof colValue === 'boolean') {
            userEnterRowValues.push(colValue);
          } else if (colValue instanceof Date) {
            userEnterRowValues.push(moment(colValue).format('YYYY/MM/DD'));
          } else if (typeof colValue === 'number') {
            userEnterRowValues.push(colValue);
          } else if (colValue == null) {
            userEnterRowValues.push('');
          }
        }
        userEnterRangeValues.push(userEnterRowValues);
      }
      userEnterValues.set(range, userEnterRangeValues);
    });
    return userEnterValues;
  }
}
